### PR TITLE
[doc] dev-notes: Members should be initialized

### PR DESCRIPTION
**Pull request by maflcko**
Also, remove mention of threads that were removed long ago.

Motivation:
Make it easier to spot bugs such as #11654 and  #12426

### CODE REVIEW

**Comment by practicalswift** [comment]
utACK fa03f777880cbe245676a3ac4d5b22d2bc93dbe9

File: src/miner.cpp [hunk 1]
```diff
@@ -30,12 +30,6 @@
 #include <queue>
 #include <utility>
 
-//////////////////////////////////////////////////////////////////////////////
-//
-// BitcoinMiner
-//
-
-//
```
Comment (jamesob) [inline src/miner.cpp]: Accidentally committed?

**Comment by jamesob** [comment]
ACK https://github.com/bitcoin/bitcoin/pull/12434/commits/fa03f777880cbe245676a3ac4d5b22d2bc93dbe9

**Review by dcousens — APPROVED** [review]

**Comment by promag** [comment]
ACK fa03f77.

File: doc/developer-notes.md [hunk 2]
```diff
@@ -380,6 +376,17 @@ C++ data structures
   - *Rationale*: Easier to understand what is happening, thus easier to spot mistakes, even for those
   that are not language lawyers
 
+- Initialize all non-static class members where they are defined
+
+  - *Rationale*: Initializing the members in the declaration makes it easy to spot uninitialized ones,
+  and avoids accidentally reading uninitialized memory
+
+```cpp
+struct A {
```
Comment (laanwj) [inline doc/developer-notes.md]: You're talking about class members above, but using a struct here. I think there can be good (performance) arguments not to have an auto-generated constructor for a POD struct, so might be better to use a class for illustration.

Comment (MarcoFalke) [reply doc/developer-notes.md]: My opinion is that such cases should just be justified with a comment:
```cpp
int a; // Not initialized for performance reasons.

Comment (MarcoFalke) [reply doc/developer-notes.md]: Anyway, fixed...

**Comment by fanquake** [comment]
utACK fa94614

**Pull request merged**